pkg/datasource/formatters/json/json.go

```diff
@@ -0,0 +1,333 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package json
+
+import (
+	"fmt"
+	"math"
+	"strconv"
+	"unicode/utf8"
+	_ "unsafe"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+)
+
+type Formatter struct {
+	ds                datasource.DataSource
+	fns               []func(e *encodeState, data datasource.Data)
+	fields            []string
+	showFields        map[string]struct{}
+	hideFields        map[string]struct{}
+	allRelativeFields bool
+	useDefault        bool
+}
+
+func New(ds datasource.DataSource, options ...Option) *Formatter {
+	f := &Formatter{
+		ds:         ds,
+		showFields: map[string]struct{}{},
+		hideFields: map[string]struct{}{},
+		useDefault: true,
+	}
+	for _, o := range options {
+		o(f)
+	}
+	f.init()
+	return f
+}
+
+func (f *Formatter) init() {
+	for _, field := range f.fields {
+		if len(field) == 0 {
+			continue
+		}
+		switch field[0] {
+		case '+':
+			f.showFields[field[1:]] = struct{}{}
+		case '-':
+			f.hideFields[field[1:]] = struct{}{}
+		default:
+			f.showFields[field] = struct{}{}
+			f.allRelativeFields = false
+		}
+	}
+
+	f.fns = append(f.fns, func(e *encodeState, data datasource.Data) {
+		e.Write([]byte("{"))
+	})
+	subFieldFuncs, _ := f.addSubFields(nil, "")
+	f.fns = append(f.fns, subFieldFuncs...)
+	f.fns = append(f.fns, func(e *encodeState, data datasource.Data) {
+		e.Write([]byte("}"))
+	})
+}
+
+func (f *Formatter) addSubFields(accessors []datasource.FieldAccessor, prefix string) (fns []func(*encodeState, datasource.Data), fieldCounter int) {
+	if accessors == nil {
+		accessors = f.ds.Accessors(true)
+	}
+
+	ctr := -1
+	for _, acc := range accessors {
+		accessor := acc
+		fullFieldName := prefix + accessor.Name()
+
+		var subFieldFuncs []func(state *encodeState, data datasource.Data)
+		var subFieldCount int
+		subFields := accessor.SubFields()
+		if len(subFields) > 0 {
+			subFieldFuncs, subFieldCount = f.addSubFields(subFields, fullFieldName+".")
+			fieldCounter += subFieldCount
+		}
+
+		// If subFieldCount is > 0, a child of this field has been requested, so we also
+		// need to show this parent; if not, we follow the default rules of field visibility
```

Comment:
No. When using the JSON format, all the fields must be visible.

Reply:
Why? I see this as a feature, not a bug - you can request exactly what you need. Maybe we should add a field selector for `*`, but I'd really like to keep this.

Reply:
I still disagree regarding this being a bug 😄 I think it's always better to give the user options.
This also has potentially a huge impact on bandwidth and CPU usage: Don't need the container image name? Don't ask for it. Lots of bytes saved, especially on huge clusters. I'd always prefer to have an option to tailor the output to exactly what you need to save resources - especially when thinking about IG as a lightweight observability tool.
`jq` might be a great tool on the client, but we have to think about client/server and scalability as well.
I hope this convinces you that this is a good feature to have.

As to what the _default_ is: I'd still prefer to use what the gadget + operators set as visible to be the default in general. As a gadget author you also have the ability to show/hide fields using annotations to adjust the default visibility - and with a (really simple) future PR maybe even specific to the output mode (JSON/columns/...). As the user you can simply change that (the relative +-%fieldname% will still work). A list of the fields should be available from the help command with the next push.

But if there's consensus that the default really _has_ to be "all", then I'll of course change it.

Reply:
> > I still disagree regarding this being a bug 😄 I think it's always better to give the user options. This also has potentially a huge impact on bandwidth and CPU usage: Don't need the container image name? Don't ask for it. Lots of bytes saved, especially on huge clusters. I'd always prefer to have an option to tailor the output to exactly what you need to save resources - especially when thinking about IG as a lightweight observability tool. `jq` might be a great tool on the client, but we have to think about client/server and scalability as well. I hope this convinces you that this is a good feature to have.
> 
> Scalability is indeed something I really want to keep high, so we are on the same wave length. Nonetheless, I stick to my guns on the fact we should focus on having Inspektor Gadget as one tool which does one thing well like the UNIX philosophy (you do not have filter in `ls`, you just use `grep`).
> 

Yeah, in general I agree - but that philosophy only works so far. I wouldn't want database software to handle its filtering using `grep` :) IG is complex software and to do our things _well_, sometimes means we have to do it on our own. 😄 

> > As to what the _default_ is: I'd still prefer to use what the gadget + operators set as visible to be the default in general. As a gadget author you also have the ability to show/hide fields using annotations to adjust the default visibility - and with a (really simple) future PR maybe even specific to the output mode (JSON/columns/...). As the user you can simply change that (the relative +-%fieldname% will still work). A list of the fields should be available from the help command with the next push.
> 
> How would user discover all the fields they can get? By reading the metadata file? I would like to avoid having plenty of people asking us why we miss this field from JSON and how to get it.

`--help` should do the trick (and later on maybe a special command like `ig info imagename` (as suggested by @alban)). Looking at the metadata file would actually not be enough, since it wouldn't tell you about the fields that operators add/alter.

Reply:
I don't think that's a bug. I think it's a great feature to have but shouldn't be used by default after this PR. I suggest keeping the same default behaviour and open a new PR/issue to discuss whether we want to change the default behaviour.